lib/pci.py

```diff
@@ -0,0 +1,480 @@
+            if pci_address in os.readlink(os.path.join(pci_class_path,
+                                                       interface))]
+
+
+def get_pci_class_name(pci_address):
+    """
+    Gets pci class name for given pci bus address
+
+    e.g: >>> pci.get_pci_class_name("0000:01:00.0")
+             'scsi_host'
+
+    :param pci_address: Any segment of a PCI address(1f, 0000:00:if, ...)
+
+    :return: class name for corresponding pci bus address
+    """
+    pci_class_dic = {'0104': 'scsi_host', '0c04': 'scsi_host', '0280': 'net',
+                     '0c03': 'scsi_host', '0200': 'net', '0108': 'nvme',
+                     '0106': 'ata_port'}
+    pci_class_id = get_pci_prop(pci_address, "Class")
+    if pci_class_id not in pci_class_dic:
+        return ""
+    return pci_class_dic.get(pci_class_id)
+
+
+def get_pci_type(pci_address):
+    """
+    Gets pci type for given pci bus address
+
+    e.g: >>> pci.get_pci_class_name("0000:01:00.0")
+             'fc'
+
+    :param pci_address: Any segment of a PCI address(1f, 0000:00:if, ...)
+
+    :return: type for corresponding pci bus address
+    """
+    pci_class_dic = {'0104': 'raid', '0c04': 'fc', '0280': 'ib',
+                     '0c03': 'usb', '0200': 'network', '0108': 'nvme'}
+    pci_class_id = get_pci_prop(pci_address, "Class")
+    if pci_class_id not in pci_class_dic:
+        return ""
+    return pci_class_dic.get(pci_class_id)
+
+
+def get_firmware(pci_address):
+    """
+    Gets firmware of a pci_address
+
+    :param pci_address: PCI address(0000:00:if.0, ...)
+
+    :return: firmware for its interface
+    """
+    class_name = get_pci_class_name(pci_address)
+    interface = get_interfaces_in_pci_address(pci_address, class_name)
+    firmware = ''
+    if not interface:
+        return firmware
+    interface = interface[0]
+    if class_name == 'net':
+        for line in commands.getoutput('ethtool -i %s' % interface).splitlines():
+            if 'firmware-version' in line:
+                firmware = line.split()[1]
+                break
+    else:
+        for name in ['firmware_rev', 'fwrev', 'fw_version']:
+            filename = "/sys/class/%s/%s/%s" % (class_name, interface, name)
+            if os.path.isfile(filename):
+                with open(filename, 'r') as fw_file:
+                    firmware = fw_file.read().strip('\t\r\n\0').split()[0]
+                    firmware = firmware.strip(',')
+    return firmware
+
+
+def get_pci_fun_list(pci_address):
+    """
+    Gets list of functions in the given PCI address.
+    Example: in address 0000:03:00, functions are 0000:03:00.0 and 0000:03:00.1
+
+    :param pci_address: Any segment of a PCI address (1f, 0000:00:1f, ...)
+
+    :return: list of functions in a PCI address.
+    """
+    return list(dev for dev in get_pci_addresses() if pci_address in dev)
+
+
+def get_slot_from_sysfs(full_pci_address):
+    """
+    Gets the PCI slot of given address.
+
+    :note: Specific for ppc64 processor.
```

Comment:
instead we can check for platform and warn in the code if not.

Reply:
will add as TODO, and address in a separate patch

Reply:
not seeing `todo` in code, pls add, rest of the changes LGTM.

Reply:
@sathnaga function now returns empty string if not ppc64